### PR TITLE
feat: webhook creation cleanup and health checks- RK-19271 

### PR DESCRIPTION
**Pull request by gosharo**
Signed-off-by: Gosha <gosha@rookout.com>

### Description

This PR will give the following features:
1. health check of the webhooks
2. OPT IN cleanup. there are race condition, but will be used only for e2e tests.
2.1. we have un-used recoverHook method that can be implemented in future for reconcile loop and integrated to diagnosis method.
3. e2e test OPT IN the GIT_WEBHOOK_AUTO_CLEANUP env var
### Related Issue(s)

Fixes: #93 
Fixes:  #105

### Checklist

Before submitting this pull request, please ensure that you have completed the following tasks:

- [x] Reviewed the [Contributing Guidelines](../docs/CONTRIBUTING.md) for the Piper project.
- [x] Ensured that your changes follow the [coding guidelines and style](../docs/CONTRIBUTING.md#coding-guidelines) of the project.
- [x] Run the tests locally and made sure they pass.
- [x] Updated the relevant documentation, if applicable.

### CODE REVIEW

File: pkg/webhook_creator/main.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+			}
+			wc.hooks[i] = newHook
+			return nil
+		}
+	}
+	return fmt.Errorf("unable to recover hookID %d, not found in list of hooks", hookID)
+}
+
+func (wc *WebhookCreatorImpl) SetHealth(status bool, hookID *int64) error {
+	//hook, ok := wc.hooks[*hookID]
```
Comment (gosharo) [inline pkg/webhook_creator/main.go:54]: this one, not working...
not sure why

File: pkg/server/routes/healthz.go [hunk 2]
```diff
@@ -1,15 +1,28 @@
...
 
-func AddHealthRoutes(rg *gin.RouterGroup) {
+func AddHealthRoutes(rg *gin.RouterGroup, wc *webhook_creator.WebhookCreatorImpl) {
 	health := rg.Group("/healthz")
 
 	health.GET("", func(c *gin.Context) {
+		ctx := c.Copy().Request.Context()
+		ctx2, cancel := context.WithTimeout(ctx, 5*time.Second)
+		defer cancel()
+		err := wc.RunDiagnosis(&ctx2)
```
Comment (qadiludmer) [inline pkg/server/routes/healthz.go:20]: Do you know how much time the diagnosis should take ?

What happens if it takes more then the health check period ? 

The pod will be killed during the diagnosis ?

File: pkg/webhook_creator/main.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	"golang.org/x/net/context"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type WebhookCreatorImpl struct {
+	clients *clients.Clients
+	cfg     *conf.GlobalConfig
+	hooks   map[int64]*git_provider.HookWithStatus
```
Comment (qadiludmer) [inline pkg/webhook_creator/main.go:18]: you access this map from 3 different go routines which might race condition each other. 

consider using a lock which iterating over the map so it won't get modified by another go routine durning the iteration

File: pkg/webhook_creator/main.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	if err != nil {
+		log.Printf("Failed to unset webhooks, error: %v", err)
+	}
+}
+
+func (wc *WebhookCreatorImpl) RunDiagnosis(ctx *context.Context) error {
+	log.Printf("Starting webhook diagnostics")
+	wc.setAllHooksHealth(false)
+	wc.pingHooks(ctx)
+	wc.checkHooksHealth(10 * time.Second)
```
Comment (qadiludmer) [inline pkg/webhook_creator/main.go:110]: you are not checking the return value of checkHooksHealth

File: pkg/webhook_creator/main.go (hunk 1 above)
Comment (qadiludmer) [reply pkg/webhook_creator/main.go:54]: What error did you get ?

Comment (qadiludmer) [reply pkg/webhook_creator/main.go:54]: this looks like the proper way to access a map

**Review by qadiludmer — COMMENTED** [review]
see comments

File: pkg/webhook_creator/main.go (hunk 1 above)
Comment (gosharo) [reply pkg/webhook_creator/main.go:54]: the key missing..

File: pkg/server/routes/healthz.go (hunk 2 above)
Comment (gosharo) [reply pkg/server/routes/healthz.go:20]:
```          livenessProbe:
            httpGet:
              path: /healthz
              port: 8080
              scheme: HTTP
            initialDelaySeconds: 10
            timeoutSeconds: 10
            periodSeconds: 60
            successThreshold: 1
            failureThreshold: 4```

File: pkg/event_handler/github_event_notifier_test.go [hunk 5]
```diff
@@ -44,6 +44,14 @@ func (m *mockGitProvider) SetStatus(ctx *context.Context, repo *string, commit *
 	return nil
 }
 
+func (m *mockGitProvider) PingHook(ctx *context.Context, hook *git_provider.HookWithStatus) error {
```
Comment (EliRookout) [inline pkg/event_handler/github_event_notifier_test.go:47]: Why do we need hook argument?

File: pkg/webhook_creator/main.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+type WebhookCreatorImpl struct {
+	clients *clients.Clients
+	cfg     *conf.GlobalConfig
+	hooks   map[int64]*git_provider.HookWithStatus
```
Comment (qadiludmer) [inline pkg/webhook_creator/main.go:19]: why do you use pointers here ?

File: pkg/git_provider/types.go [hunk 7]
```diff
@@ -5,6 +5,12 @@ import (
 	"net/http"
 )
 
+type HookWithStatus struct {
+	HookID       *int64
```
Comment (qadiludmer) [inline pkg/git_provider/types.go:9]: why pointer ?

File: pkg/git_provider/types.go [hunk 8]
```diff
@@ -5,6 +5,12 @@ import (
...
+	HealthStatus bool
+	RepoName     *string
```
Comment (qadiludmer) [inline pkg/git_provider/types.go:11]: why pointer ?

File: pkg/server/routes/healthz.go [hunk 9]
```diff
@@ -1,15 +1,28 @@
...
 
 	"github.com/gin-gonic/gin"
 )
 
-func AddHealthRoutes(rg *gin.RouterGroup) {
+func AddHealthRoutes(rg *gin.RouterGroup, wc *webhook_creator.WebhookCreatorImpl) {
 	health := rg.Group("/healthz")
 
 	health.GET("", func(c *gin.Context) {
+		ctx := c.Copy().Request.Context()
```
Comment (ElDuderinos) [inline pkg/server/routes/healthz.go:17]: Do you really need to copy? I think it's only needed when passing the ctx to a different routine

File: pkg/server/routes/webhook.go [hunk 10]
```diff
@@ -45,7 +50,7 @@ func AddWebhookRoutes(cfg *conf.GlobalConfig, clients *clients.Clients, rg *gin.
 			err = clients.Workflows.HandleWorkflowBatch(&ctx, wf)
 			if err != nil {
 				c.AbortWithStatusJSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
-				log.Printf("failed to handle workflow, error: %v", err)
+				log.Printf("failed to handle workflow, error: %v", err) //Muli
```
Comment (ElDuderinos) [inline pkg/server/routes/webhook.go:53]: Muli?

File: pkg/server/main.go [hunk 11]
```diff
@@ -12,13 +12,9 @@ func Start(ctx context.Context, stop context.CancelFunc, cfg *conf.GlobalConfig,
 	srv := NewServer(cfg, clients)
 	gracefulShutdownHandler := NewGracefulShutdown(ctx, stop)
 	httpServer := srv.ListenAndServe()
+	srv.webhookCreator.Start()
```
Comment (ElDuderinos) [inline pkg/server/main.go:15]: Why aren't you implicitly starting the webhook creator upon server start?
Is there any scenario in which the server starts and you don't want to start the webhook creator?
More over, if you're starting it out of the server here, then I'd use a function srv.StartWebhookCreator() or maybe srv.StartServices() and put the webhookCreator.Start() there and other services in the future.
TL;DR - no sense in having the main know the webhook creator imo.

File: pkg/server/shutdown.go [hunk 12]
```diff
@@ -33,12 +33,9 @@ func (s *gracefulShutdown) Shutdown(httpServer *http.Server, clients *clients.Cl
 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
 	defer cancel()
 
-	err := clients.GitProvider.UnsetWebhook(&ctx)
-	if err != nil {
-		log.Println("Unset webhook error: ", err) // ERROR
-	}
+	webhookCreator.Stop(&ctx)
```
Comment (ElDuderinos) [inline pkg/server/shutdown.go:36]: Same comment for the start in the main, I'd go with ShutdownServices for the server.

File: pkg/utils/common.go [hunk 13]
```diff
@@ -122,6 +122,10 @@ func BPtr(b bool) *bool {
 	return &b
 }
 
+func IPtr(i int64) *int64 {
+	return &i
+}
+
```
Comment (ElDuderinos) [inline pkg/utils/common.go:128]: Comment relevant for the BPtr and SPtr.
Why do you need these? why would you need a wrapper for & ? 
:)

File: pkg/webhook_creator/main.go [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+
+		if allHealthy {
+			return true
+		}
+
+		if time.Since(startTime) >= timeout {
+			break
+		}
+
+		time.Sleep(1 * time.Second) // Adjust the sleep duration as per your requirement
```
Comment (ElDuderinos) [inline pkg/webhook_creator/main.go:142]: You anyway have a cfg here, why not config it and get it from the cfg?

File: pkg/webhook_creator/main.go [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+func (wc *WebhookCreatorImpl) RunDiagnosis(ctx *context.Context) error {
+	log.Printf("Starting webhook diagnostics")
+	wc.setAllHooksHealth(false)
+	err := wc.pingHooks(ctx)
+	if err != nil {
+		return err
+	}
+	if !wc.checkHooksHealth(5 * time.Second) {
```
Comment (ElDuderinos) [inline pkg/webhook_creator/main.go:183]: use cfg for this maybe?

**Review by ElDuderinos — APPROVED** [review]
